src/fdb5/remote/client/ClientConnection.cc

```diff
@@ -0,0 +1,483 @@
+        std::stringstream ss;
+        ss << "Session ID does not match session received from server: "
+           << sessionID_ << " != " << clientSession;
+        throw eckit::BadValue(ss.str(), Here());
+    }
+    if (serverFunctionality.has("NumberOfConnections") && serverFunctionality.getInt("NumberOfConnections")==1) {
+        single_ = true;
+    }
+
+    if (single_ && !(dataEndpoint_ == controlEndpoint_)) {
+        eckit::Log::warning() << "Returned control interface does not match. "
+                       << dataEndpoint_ << " /= "
+                       << controlEndpoint_ << std::endl;
+    }
+
+    return serverSession;
+}
+
+void ClientConnection::listeningControlThreadLoop() {
+
+    try {
+
+        MessageHeader hdr;
+
+        while (true) {
+
+            eckit::Buffer payload = Connection::readControl(hdr);
+
+            LOG_DEBUG_LIB(LibFdb5) << "ClientConnection::listeningControlThreadLoop - got [message=" << hdr.message << ",clientID=" << hdr.clientID() << ",control=" << hdr.control() << ",requestID=" << hdr.requestID << ",payload=" << hdr.payloadSize << "]" << std::endl;
+
+            if (hdr.message == Message::Exit) {
+                controlStopping_ = true;
+                return;
+            } else {
+                if (hdr.clientID()) {
+                    bool handled = false;
+
+                    ASSERT(hdr.control() || single_);
+
+                    auto pp = promises_.find(hdr.requestID);
+                    if (pp != promises_.end()) {
+                        std::lock_guard<std::mutex> lock(promisesMutex_);
+                        if (hdr.payloadSize == 0) {
+                            ASSERT(hdr.message == Message::Received);
+                            pp->second.set_value(eckit::Buffer(0));
+                        } else {
+                            pp->second.set_value(std::move(payload));
+                        }
+                        promises_.erase(pp);
+                        handled = true;
+                    } else {
+                        Client* client = nullptr;
+                        {
+                            std::lock_guard<std::mutex> lock(clientsMutex_);
+
+                            auto it = clients_.find(hdr.clientID());
+                            if (it == clients_.end()) {
+                                std::stringstream ss;
+                                ss << "ERROR: connection=" << controlEndpoint_ << " received [clientID="<< hdr.clientID() << ",requestID="<< hdr.requestID << ",message=" << hdr.message << ",payload=" << hdr.payloadSize << "]" << std::endl;
+                                ss << "Unexpected answer for clientID recieved (" << hdr.clientID() << "). ABORTING";
+                                eckit::Log::status() << ss.str() << std::endl;
+                                eckit::Log::error() << "Retrieving... " << ss.str() << std::endl;
+                                throw eckit::SeriousBug(ss.str(), Here());
```

Comment:
At the end of this code block it is considered to be a serious bug if the client id is unknown, however we only enter this code if there is ANY client ID. If there is no client ID the code will just continue with the next iteration of the for-ever loop. Is this intentional? I would argue that no client id is the same error case as an unknown client id.